Libplanet/Net/Protocols/KademliaProtocol.cs

```diff
@@ -294,65 +297,105 @@ public string Trace()
+                        "{BoundPeer}, a target peer, is in the routing table does not respond.";
+                    _logger.Debug(msg, boundPeer);
+                    return null;
+                }
+
+                _logger.Debug(
+                    "{BoundPeer}, a target peer, is in the routing table.",
+                    boundPeer);
+                return boundPeer;
             }
 
-            IEnumerable<BoundPeer> found;
-            if (viaPeer is null)
+            var history = new ConcurrentBag<BoundPeer>();
+            var peersToFind = new ConcurrentQueue<BoundPeer>();
+            foreach (var peer in _routing.Neighbors(target, _findConcurrency, false))
             {
-                found = await QueryNeighborsAsync(history, target, timeout, cancellationToken);
+                peersToFind.Enqueue(peer);
             }
-            else
+
+            while (peersToFind.Any())
             {
-                found = await GetNeighbors(viaPeer, target, timeout, cancellationToken);
+                if (cancellationToken.IsCancellationRequested)
+                {
+                    throw new TaskCanceledException(
+                        $"Task is cancelled during {nameof(FindSpecificPeerAsync)}()");
+                }
+
+                if (!peersToFind.TryDequeue(out BoundPeer viaPeer))
+                {
+                    continue;
+                }
+
                 history.Add(viaPeer);
-            }
+                List<BoundPeer> foundPeers =
+                    (await GetNeighbors(viaPeer, target, timeout, cancellationToken)).ToList();
 
-            BoundPeer specificPeerFound = await ProcessFoundForSpecificAsync(
-                history,
-                found,
-                target,
-                depth,
-                timeout,
-                cancellationToken,
-                searchTarget
-            );
+                int count = 0;
+                foreach (var found in foundPeers
+                    .Where(peer => !history.Contains(peer) && !peer.Address.Equals(_address))
+                    .Take(_findConcurrency))
+                {
+                    try
+                    {
+                        await PingAsync(found, _requestTimeout, cancellationToken);
+                        if (found.Address.Equals(target))
+                        {
+                            return found;
+                        }
+
+                        peersToFind.Enqueue(found);
+                        if (count++ >= _findConcurrency)
+                        {
+                            break;
+                        }
+                    }
+                    catch (TaskCanceledException)
+                    {
+                        throw new TaskCanceledException(
+                            $"Task is cancelled during {nameof(FindSpecificPeerAsync)}()");
+                    }
+                    catch (Exception)
+                    {
+                        continue;
```

Comment:
is it okay we ignore here without any logging?

Reply:
I guess should ignore `PingTimeoutException` only, otherwise just throw. I'll fix it.